modules/c/nrt/source/IOInterface.c

```diff
@@ -190,19 +190,21 @@ NRTPRIV(NRT_BOOL) BufferAdapter_read(NRT_DATA * data, void *buf, size_t size,
                                      nrt_Error * error)
 {
     BufferIOControl *control = (BufferIOControl *) data;
+    const size_t validBufferSize = control->mark >= control->size ? 0 :
+        control->size - control->mark;
 
-    if (size > control->size - control->mark)
+    if (validBufferSize < size)
     {
-        nrt_Error_init(error, "Invalid size requested - EOF", NRT_CTXT,
-                       NRT_ERR_MEMORY);
-        return NRT_FAILURE;
+        memcpy(buf, (char *)(control->buf + control->mark), validBufferSize);
```

Comment:
This interface is weird in that usually if short reads are possible, there's some way to tell the caller you read less than they asked for, but the interface doesn't provide a way to do that... if I look at `nrt_IOHandle_read()`, in both the Unix and Windows implementations, if we do a `read()` and get an `EOF`, we error out.  So how is the BufferAdapter interface not doing an equivalent thing right now (i.e. is the change you're putting in here really consistent)?